File: ios/XiriGoEcommerce/Core/DesignSystem/Component/XGLoadingView.swift
import SwiftUI

// MARK: - DefaultSkeletonConstants

/// Internal constants for default skeleton placeholders.
private enum DefaultSkeletonConstants {
    static let boxHeight: CGFloat = 200
    static let lineHeight: CGFloat = 14
    static let inlineLineHeight: CGFloat = 24
    static let longLineFraction: CGFloat = 0.7
    static let shortLineFraction: CGFloat = 0.5
    static let lineBlockRowCount: CGFloat = 3
}

// MARK: - XGLoadingView

/// Skeleton-aware full-screen loading view.
///
/// Crossfades between a skeleton placeholder and real content using
/// ``XGMotion/Crossfade/contentSwitch`` (0.2s). When no skeleton slot is provided,
/// a default full-width shimmer layout is shown.
///
/// Usage with crossfade:
/// ```swift
/// XGLoadingView(isLoading: viewModel.isLoading) {
///     SkeletonBox(width: 170, height: 170)
///     SkeletonLine(width: 140)
/// } content: {
///     ProductDetailContent(product: product)
/// }
/// ```
///
/// Usage without content (loading-only, backward compatible):
/// ```swift
/// XGLoadingView()
/// ```
struct XGLoadingView<Skeleton: View, Content: View>: View {
    // MARK: - Lifecycle

    /// Creates a skeleton-aware loading view with crossfade transition.
    ///
    /// - Parameters:
    ///   - isLoading: When `true`, shows the skeleton. When `false`, shows content.
    ///   - skeleton: The skeleton placeholder to show during loading.
    ///   - content: The real content to reveal when loading completes.
    init(
        isLoading: Bool,
        @ViewBuilder skeleton: () -> Skeleton,
        @ViewBuilder content: () -> Content,
    ) {
        self.isLoading = isLoading
        self.skeleton = skeleton()
        self.content = content()
    }

    // MARK: - Internal

    let isLoading: Bool
    let skeleton: Skeleton
    let content: Content

    var body: some View {
        Group {
            if isLoading {
                skeleton
                    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
                    .transition(.opacity)
                    .accessibilityLabel(Text("skeleton_loading_placeholder"))
            } else {
                content
                    .transition(.opacity)
            }
        }
        .animation(
            .easeInOut(duration: XGMotion.Crossfade.contentSwitch),
            value: isLoading,
        )
    }
}

// MARK: - XGLoadingView convenience (no skeleton)

extension XGLoadingView where Skeleton == DefaultFullScreenSkeleton {
    /// Creates a skeleton-aware loading view with a default shimmer placeholder.
    ///
    /// - Parameters:
    ///   - isLoading: When `true`, shows default shimmer. When `false`, shows content.
    ///   - content: The real content to reveal when loading completes.
    init(
        isLoading: Bool,
        @ViewBuilder content: () -> Content,
    ) {
        self.isLoading = isLoading
        skeleton = DefaultFullScreenSkeleton()
        self.content = content()
    }
}

extension XGLoadingView where Skeleton == DefaultFullScreenSkeleton, Content == EmptyView {
    /// Creates a loading-only view with the default shimmer placeholder (no content slot).
    ///
    /// Use this for backward compatibility where the caller switches on UI state
    /// and renders content in a different branch.
    init() {
        isLoading = true
        skeleton = DefaultFullScreenSkeleton()
        content = EmptyView()
    }
}

// MARK: - XGLoadingView convenience (custom skeleton, no content)

extension XGLoadingView where Content == EmptyView {
    /// Creates a loading-only view with a custom skeleton placeholder.
    ///
    /// - Parameter skeleton: The skeleton placeholder to display.
    init(@ViewBuilder skeleton: () -> Skeleton) {
        isLoading = true
        self.skeleton = skeleton()
        content = EmptyView()
    }
}

// MARK: - XGLoadingIndicator

/// Skeleton-aware inline loading indicator for list footers and sections.
///
/// Crossfades between a skeleton placeholder and real content using
/// ``XGMotion/Crossfade/contentSwitch`` (0.2s). When no skeleton slot is provided,
/// a default full-width shimmer line is shown.
struct XGLoadingIndicator<Skeleton: View, Content: View>: View {
    // MARK: - Lifecycle

    /// Creates a skeleton-aware inline loading indicator with crossfade transition.
    ///
    /// - Parameters:
    ///   - isLoading: When `true`, shows the skeleton. When `false`, shows content.
    ///   - skeleton: The skeleton placeholder to show during loading.
    ///   - content: The real content to reveal when loading completes.
    init(
        isLoading: Bool,
        @ViewBuilder skeleton: () -> Skeleton,
        @ViewBuilder content: () -> Content,
    ) {
        self.isLoading = isLoading
        self.skeleton = skeleton()
        self.content = content()
    }

    // MARK: - Internal

    let isLoading: Bool
    let skeleton: Skeleton
    let content: Content

    var body: some View {
        Group {
            if isLoading {
                skeleton
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, XGSpacing.base)
                    .transition(.opacity)
                    .accessibilityLabel(Text("skeleton_loading_placeholder"))
            } else {
                content
                    .transition(.opacity)
            }
        }
        .animation(
            .easeInOut(duration: XGMotion.Crossfade.contentSwitch),
            value: isLoading,
        )
    }
}

// MARK: - XGLoadingIndicator convenience (no skeleton)

extension XGLoadingIndicator where Skeleton == DefaultInlineSkeleton {
    /// Creates an inline loading indicator with a default shimmer line.
    ///
    /// - Parameters:
    ///   - isLoading: When `true`, shows default shimmer. When `false`, shows content.
    ///   - content: The real content to reveal when loading completes.
    init(
        isLoading: Bool,
        @ViewBuilder content: () -> Content,
    ) {
        self.isLoading = isLoading
        skeleton = DefaultInlineSkeleton()
        self.content = content()
    }
}

extension XGLoadingIndicator where Skeleton == DefaultInlineSkeleton, Content == EmptyView {
    /// Creates a loading-only inline indicator with the default shimmer line.
    init() {
        isLoading = true
        skeleton = DefaultInlineSkeleton()
        content = EmptyView()
    }
}

// MARK: - DefaultFullScreenSkeleton

/// Default full-screen shimmer placeholder: a shimmer box with shimmer lines below.
struct DefaultFullScreenSkeleton: View {
    var body: some View {
        VStack(alignment: .leading, spacing: XGSpacing.sm) {
            RoundedRectangle(cornerRadius: XGCornerRadius.medium)
                .fill(XGColors.shimmer)
                .frame(height: DefaultSkeletonConstants.boxHeight)
                .shimmerEffect()

            GeometryReader { geometry in
                VStack(alignment: .leading, spacing: XGSpacing.sm) {
                    RoundedRectangle(cornerRadius: XGCornerRadius.small)
                        .fill(XGColors.shimmer)
                        .frame(
                            width: geometry.size.width * DefaultSkeletonConstants.longLineFraction,
                            height: DefaultSkeletonConstants.lineHeight,
                        )
                        .shimmerEffect()

                    RoundedRectangle(cornerRadius: XGCornerRadius.small)
                        .fill(XGColors.shimmer)
                        .frame(
                            width: geometry.size.width * DefaultSkeletonConstants.shortLineFraction,
                            height: DefaultSkeletonConstants.lineHeight,
                        )
                        .shimmerEffect()
                }
            }
            .frame(height: DefaultSkeletonConstants.lineHeight * DefaultSkeletonConstants.lineBlockRowCount)
        }
        .padding(XGSpacing.base)
    }
}

// MARK: - DefaultInlineSkeleton

/// Default inline shimmer placeholder: a full-width shimmer line.
struct DefaultInlineSkeleton: View {
    var body: some View {
        RoundedRectangle(cornerRadius: XGCornerRadius.small)
            .fill(XGColors.shimmer)
            .frame(maxWidth: .infinity)
            .frame(height: DefaultSkeletonConstants.inlineLineHeight)
            .shimmerEffect()
    }
}

// MARK: - PreviewConstants

private enum PreviewConstants {
    static let skeletonBoxSize: CGFloat = 170
    static let skeletonLineWidthLong: CGFloat = 140
    static let skeletonLineWidthShort: CGFloat = 80
    static let indicatorLineWidth: CGFloat = 200
}

// MARK: - Previews

#Preview("XGLoadingView — Default Skeleton") {
    XGLoadingView()
        .xgTheme()
}

#Preview("XGLoadingView — Custom Skeleton") {
    XGLoadingView {
        VStack(alignment: .leading, spacing: XGSpacing.sm) {
            SkeletonBox(
                width: PreviewConstants.skeletonBoxSize,
                height: PreviewConstants.skeletonBoxSize,
            )
            SkeletonLine(width: PreviewConstants.skeletonLineWidthLong)
            SkeletonLine(width: PreviewConstants.skeletonLineWidthShort)
        }
        .padding(XGSpacing.base)
    }
    .xgTheme()
}

#Preview("XGLoadingView — Crossfade Loading") {
    XGLoadingView(isLoading: true) {
        VStack(alignment: .leading, spacing: XGSpacing.sm) {
            SkeletonBox(
                width: PreviewConstants.skeletonBoxSize,
                height: PreviewConstants.skeletonBoxSize,
            )
            SkeletonLine(width: PreviewConstants.skeletonLineWidthLong)
        }
        .padding(XGSpacing.base)
    } content: {
        Text(verbatim: "Content loaded!")
            .padding(XGSpacing.base)
    }
    .xgTheme()
}

#Preview("XGLoadingView — Crossfade Content") {
    XGLoadingView(isLoading: false) {
        Text(verbatim: "Content loaded!")
            .padding(XGSpacing.base)
    }
    .xgTheme()
}

#Preview("XGLoadingIndicator — Default Skeleton") {
    VStack {
        Text(verbatim: "List content above")
        XGLoadingIndicator()
    }
    .xgTheme()
}

#Preview("XGLoadingIndicator — Custom Skeleton") {
    VStack {
        Text(verbatim: "List content above")
        XGLoadingIndicator(isLoading: true) {
            SkeletonLine(width: PreviewConstants.indicatorLineWidth)
        } content: {
            Text(verbatim: "More items loaded")
        }
    }
    .xgTheme()
}

// MARK: - XGLoadingViewInteractiveDemo

/// Interactive demo showing the crossfade transition between skeleton and content.
private struct XGLoadingViewInteractiveDemo: View {
    // MARK: - Internal

    var body: some View {
        VStack {
            XGLoadingView(isLoading: isLoading) {
                VStack(alignment: .leading, spacing: XGSpacing.sm) {
                    SkeletonBox(
                        width: PreviewConstants.skeletonBoxSize,
                        height: PreviewConstants.skeletonBoxSize,
                    )
                    SkeletonLine(width: PreviewConstants.skeletonLineWidthLong)
                    SkeletonLine(width: PreviewConstants.skeletonLineWidthShort)
                }
                .padding(XGSpacing.base)
            } content: {
                Text(verbatim: "Real content loaded!")
                    .padding(XGSpacing.base)
            }

            Button {
                isLoading.toggle()
            } label: {
                Text(verbatim: isLoading ? "Show Content" : "Show Skeleton")
            }
            .padding(XGSpacing.base)
        }
    }

    // MARK: - Private

    @State private var isLoading = true
}

#Preview("XGLoadingView — Interactive Toggle") {
    XGLoadingViewInteractiveDemo()
        .xgTheme()
}

File: android/app/src/main/java/com/xirigo/ecommerce/core/designsystem/component/XGLoadingView.kt
package com.xirigo.ecommerce.core.designsystem.component

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.xirigo.ecommerce.R
import com.xirigo.ecommerce.core.designsystem.theme.XGColors
import com.xirigo.ecommerce.core.designsystem.theme.XGCornerRadius
import com.xirigo.ecommerce.core.designsystem.theme.XGMotion
import com.xirigo.ecommerce.core.designsystem.theme.XGSpacing
import com.xirigo.ecommerce.core.designsystem.theme.XGTheme

// Default shimmer placeholder heights
private val DefaultShimmerBoxHeight = 200.dp
private val DefaultShimmerLineHeight = 14.dp
private val InlineShimmerLineHeight = 24.dp

/**
 * Skeleton-aware full-screen loading view.
 *
 * Crossfades between a skeleton placeholder and real content using
 * [XGMotion.Crossfade.CONTENT_SWITCH] (200 ms). When no [skeleton] slot is provided,
 * a default full-width shimmer box is shown.
 *
 * @param isLoading When `true`, shows the skeleton/shimmer. When `false`, shows [content].
 * @param modifier Modifier applied to the outer container.
 * @param skeleton Optional skeleton placeholder slot. When `null`, a default shimmer box is used.
 * @param content The real content to reveal when loading completes.
 */
@Composable
fun XGLoadingView(
    isLoading: Boolean,
    modifier: Modifier = Modifier,
    skeleton: (@Composable () -> Unit)? = null,
    content: @Composable () -> Unit,
) {
    val loadingDescription = stringResource(R.string.skeleton_loading_placeholder)

    AnimatedContent(
        targetState = isLoading,
        modifier = modifier.fillMaxSize(),
        transitionSpec = {
            fadeIn(tween(XGMotion.Crossfade.CONTENT_SWITCH)) togetherWith
                fadeOut(tween(XGMotion.Crossfade.CONTENT_SWITCH))
        },
        label = "xgLoadingViewCrossfade",
    ) { loading ->
        if (loading) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .semantics { contentDescription = loadingDescription },
            ) {
                if (skeleton != null) {
                    skeleton()
                } else {
                    DefaultFullScreenSkeleton()
                }
            }
        } else {
            content()
        }
    }
}

/**
 * Convenience overload that always shows the loading skeleton (no content slot).
 *
 * Use this when the caller only needs a loading placeholder without content transition,
 * for example inside a `when (uiState)` block where content is rendered by a different branch.
 *
 * @param modifier Modifier applied to the outer container.
 * @param skeleton Optional skeleton placeholder slot. When `null`, a default shimmer box is used.
 */
@Composable
fun XGLoadingView(modifier: Modifier = Modifier, skeleton: (@Composable () -> Unit)? = null) {
    val loadingDescription = stringResource(R.string.skeleton_loading_placeholder)

    Box(
        modifier = modifier
            .fillMaxSize()
            .semantics { contentDescription = loadingDescription },
    ) {
        if (skeleton != null) {
            skeleton()
        } else {
            DefaultFullScreenSkeleton()
        }
    }
}

/**
 * Skeleton-aware inline loading indicator for list footers and sections.
 *
 * Crossfades between a skeleton placeholder and real content using
 * [XGMotion.Crossfade.CONTENT_SWITCH] (200 ms). When no [skeleton] slot is provided,
 * a default full-width shimmer line is shown.
 *
 * @param isLoading When `true`, shows the skeleton/shimmer. When `false`, shows [content].
 * @param modifier Modifier applied to the outer container.
 * @param skeleton Optional skeleton placeholder slot. When `null`, a default shimmer line is used.
 * @param content The real content to reveal when loading completes.
 */
@Composable
fun XGLoadingIndicator(
    isLoading: Boolean,
    modifier: Modifier = Modifier,
    skeleton: (@Composable () -> Unit)? = null,
    content: @Composable () -> Unit,
) {
    val loadingDescription = stringResource(R.string.skeleton_loading_placeholder)

    AnimatedContent(
        targetState = isLoading,
        modifier = modifier.fillMaxWidth(),
        transitionSpec = {
            fadeIn(tween(XGMotion.Crossfade.CONTENT_SWITCH)) togetherWith
                fadeOut(tween(XGMotion.Crossfade.CONTENT_SWITCH))
        },
        label = "xgLoadingIndicatorCrossfade",
    ) { loading ->
        if (loading) {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(XGSpacing.Base)
                    .semantics { contentDescription = loadingDescription },
            ) {
                if (skeleton != null) {
                    skeleton()
                } else {
                    DefaultInlineSkeleton()
                }
            }
        } else {
            content()
        }
    }
}

/**
 * Convenience overload that always shows the inline loading skeleton (no content slot).
 *
 * @param modifier Modifier applied to the outer container.
 * @param skeleton Optional skeleton placeholder slot. When `null`, a default shimmer line is used.
 */
@Composable
fun XGLoadingIndicator(modifier: Modifier = Modifier, skeleton: (@Composable () -> Unit)? = null) {
    val loadingDescription = stringResource(R.string.skeleton_loading_placeholder)

    Box(
        modifier = modifier
            .fillMaxWidth()
            .padding(XGSpacing.Base)
            .semantics { contentDescription = loadingDescription },
    ) {
        if (skeleton != null) {
            skeleton()
        } else {
            DefaultInlineSkeleton()
        }
    }
}

/** Default full-screen shimmer placeholder: a set of shimmer boxes and lines. */
@Composable
private fun DefaultFullScreenSkeleton() {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(XGSpacing.Base),
        verticalArrangement = Arrangement.spacedBy(XGSpacing.SM),
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(DefaultShimmerBoxHeight)
                .clip(RoundedCornerShape(XGCornerRadius.Medium))
                .background(XGColors.Shimmer)
                .shimmerEffect(),
        )
        Box(
            modifier = Modifier
                .fillMaxWidth(fraction = 0.7f)
                .height(DefaultShimmerLineHeight)
                .clip(RoundedCornerShape(XGCornerRadius.Small))
                .background(XGColors.Shimmer)
                .shimmerEffect(),
        )
        Box(
            modifier = Modifier
                .fillMaxWidth(fraction = 0.5f)
                .height(DefaultShimmerLineHeight)
                .clip(RoundedCornerShape(XGCornerRadius.Small))
                .background(XGColors.Shimmer)
                .shimmerEffect(),
        )
    }
}

/** Default inline shimmer placeholder: a full-width shimmer line. */
@Composable
private fun DefaultInlineSkeleton() {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(InlineShimmerLineHeight)
            .clip(RoundedCornerShape(XGCornerRadius.Small))
            .background(XGColors.Shimmer)
            .shimmerEffect(),
    )
}

// region Preview

@Preview(showBackground = true, name = "XGLoadingView — Default Skeleton")
@Composable
private fun XGLoadingViewDefaultSkeletonPreview() {
    XGTheme {
        XGLoadingView()
    }
}

@Preview(showBackground = true, name = "XGLoadingView — Custom Skeleton")
@Composable
private fun XGLoadingViewCustomSkeletonPreview() {
    XGTheme {
        XGLoadingView(
            skeleton = {
                Column(
                    modifier = Modifier.padding(XGSpacing.Base),
                    verticalArrangement = Arrangement.spacedBy(XGSpacing.SM),
                ) {
                    SkeletonBox(width = 170.dp, height = 170.dp)
                    SkeletonLine(width = 140.dp)
                    SkeletonLine(width = 80.dp)
                }
            },
        )
    }
}

@Preview(showBackground = true, name = "XGLoadingView — Crossfade Loading")
@Composable
private fun XGLoadingViewCrossfadeLoadingPreview() {
    XGTheme {
        XGLoadingView(
            isLoading = true,
            skeleton = {
                Column(
                    modifier = Modifier.padding(XGSpacing.Base),
                    verticalArrangement = Arrangement.spacedBy(XGSpacing.SM),
                ) {
                    SkeletonBox(width = 170.dp, height = 170.dp)
                    SkeletonLine(width = 140.dp)
                }
            },
        ) {
            Text(
                text = "Content loaded!",
                style = MaterialTheme.typography.bodyLarge,
                modifier = Modifier.padding(XGSpacing.Base),
            )
        }
    }
}

@Preview(showBackground = true, name = "XGLoadingView — Crossfade Content")
@Composable
private fun XGLoadingViewCrossfadeContentPreview() {
    XGTheme {
        XGLoadingView(
            isLoading = false,
        ) {
            Text(
                text = "Content loaded!",
                style = MaterialTheme.typography.bodyLarge,
                modifier = Modifier.padding(XGSpacing.Base),
            )
        }
    }
}

@Preview(showBackground = true, name = "XGLoadingView — Interactive Toggle")
@Composable
private fun XGLoadingViewInteractivePreview() {
    XGTheme {
        var isLoading by remember { mutableStateOf(true) }

        Column {
            XGLoadingView(
                isLoading = isLoading,
                modifier = Modifier.weight(1f),
                skeleton = {
                    Column(
                        modifier = Modifier.padding(XGSpacing.Base),
                        verticalArrangement = Arrangement.spacedBy(XGSpacing.SM),
                    ) {
                        SkeletonBox(width = 170.dp, height = 170.dp)
                        SkeletonLine(width = 140.dp)
                        SkeletonLine(width = 80.dp)
                    }
                },
            ) {
                Text(
                    text = "Real content loaded!",
                    style = MaterialTheme.typography.bodyLarge,
                    modifier = Modifier.padding(XGSpacing.Base),
                )
            }

            androidx.compose.material3.Button(
                onClick = { isLoading = !isLoading },
                modifier = Modifier.padding(XGSpacing.Base),
            ) {
                Text(text = if (isLoading) "Show Content" else "Show Skeleton")
            }
        }
    }
}

@Preview(showBackground = true, name = "XGLoadingIndicator — Default Skeleton")
@Composable
private fun XGLoadingIndicatorDefaultSkeletonPreview() {
    XGTheme {
        XGLoadingIndicator()
    }
}

@Preview(showBackground = true, name = "XGLoadingIndicator — Custom Skeleton")
@Composable
private fun XGLoadingIndicatorCustomSkeletonPreview() {
    XGTheme {
        XGLoadingIndicator(
            skeleton = {
                SkeletonLine(width = 200.dp)
            },
        )
    }
}

@Preview(showBackground = true, name = "XGLoadingIndicator — Crossfade")
@Composable
private fun XGLoadingIndicatorCrossfadePreview() {
    XGTheme {
        var isLoading by remember { mutableStateOf(true) }

        Column {
            Text(text = "List content above", modifier = Modifier.padding(XGSpacing.Base))

            XGLoadingIndicator(
                isLoading = isLoading,
                skeleton = { SkeletonLine(width = 200.dp) },
            ) {
                Text(
                    text = "More items loaded",
                    modifier = Modifier.padding(XGSpacing.Base),
                )
            }

            Spacer(modifier = Modifier.height(XGSpacing.SM))

            androidx.compose.material3.Button(
                onClick = { isLoading = !isLoading },
                modifier = Modifier.padding(XGSpacing.Base),
            ) {
                Text(text = if (isLoading) "Show Content" else "Show Skeleton")
            }
        }
    }
}

// endregion
